DOC-13422-Eventing-Memory-Quota (#4055)
couchbase/docs-server: modules/learn/pages/buckets-memory-and-storage/eventing-memory-quota.adoc

> = Eventing Memory Quota 

Reviewer: ```suggestion
= Eventing Service Memory Quota 
```

Author: Done.

:description: The Eventing Service memory quota does not enforce a hard memory limit on the Eventing subsystem, including worker processes. 

> This page explains how the memory quota actually works, how Eventing distributes it across workers, and its limitations in non-containerized environments.

Reviewer: ```suggestion
This page explains how the memory quota actually works, how the Eventing Service distributes it across workers, and its limitations in non-containerized environments.
```

Author: Done.

[abstract]
{description}

> == Overview

Reviewer: ```suggestion
== About the Eventing Service Memory Quota
```

Author: Done.

> The Couchbase Eventing Service memory quota does not enforce a hard memory limit on the entire Eventing subsystem, including worker processes. 

Reviewer: ```suggestion

```
You said this already. Decide where on the page you want this information to go instead of repeating it so close on the same page.

The service uses the quota for queue sizing and JavaScript (JS) runtime memory heap, not as an absolute memory cap. 
In VM or bare-metal deployments, Eventing functions can exceed the configured quota at runtime.

> Understanding this behavior helps you to appropriately size and monitor Eventing functions in production environments.

Reviewer: ```suggestion
The Eventing Service uses the quota for:

* <<worker-queue-sizing,>>: Controls the maximum size of each worker's input queue.
* <<gc-trigger,>>: Determines when the Eventing Service invokes the JavaScript environment's GC to reclaim memory.

The memory quota is not an absolute memory cap.
The quota does not restrict the total memory consumed by JavaScript runtime heaps or Eventing processes.

The Eventing Service divides its total memory quota uniformly across all workers in all deployed functions:
....
Per-Worker quota = Total memory quota ÷ Total number of workers
....

For example, if you had a memory quota of 256{nbsp}MB, with 4 total workers, each worker would use 64{nbsp}MB of memory, or 256 MB ÷ 4 workers.

Each worker receives an equal share of the total quota, regardless of the actual memory usage of individual functions.

As you deploy more functions or add workers per function, the per-worker quota decreases proportionally.
A smaller per-worker quota can affect performance if individual functions require substantial memory for processing.

In VM or bare-metal deployments, Eventing Functions can exceed the configured quota at runtime.
Understanding this behavior helps you to appropriately size and monitor your Eventing Functions in production environments.
```

> == How Memory Quota Works

Reviewer: ```suggestion

```

> The Eventing Service memory quota controls specific aspects of memory management, such as queue sizing and frequency of garbage collection. 

Reviewer: ```suggestion

```

> However, it does not act as an absolute ceiling on memory usage.

Reviewer: ```suggestion

```

=== Minimum Size

The memory quota must be a minimum of 256 MB. 

> Couchbase does not support values lesser than 256 MB.

Reviewer: ```suggestion
The memory quota for the Eventing Service must be a minimum of 256{nbsp}MB. 
Couchbase does not support memory quota values smaller than 256{nbsp}MB.
```

=== Purpose of the Quota

The memory quota serves 2 primary purposes:

* **Worker queue sizing**: controls the maximum size of each worker's input queue.
* **Garbage Collection (GC) triggering**: determines when Eventing invokes JavaScript environment GC to reclaim memory.

> The quota does not restrict the total memory consumed by JavaScript runtime heaps or Eventing processes.

Reviewer: ```suggestion

```

> == Per-Worker Distribution

Reviewer: ```suggestion

```

> The Eventing Service divides the total memory quota uniformly across all workers in all deployed functions.

Reviewer: ```suggestion

```

=== Calculation
Eventing calculates the per-worker quota as follows:
....
Per-Worker quota = Total memory quota ÷ Total number of workers

> ....

Reviewer: ```suggestion

```

=== Example 

Consider the following configuration:

* Total memory quota: 256 MB
* Total workers (across all deployed functions): 4
* Per-worker budget: 64 MB (256 MB ÷ 4 workers)

> Each worker receives an equal share of the total quota, regardless of the actual memory usage of individual functions.

Reviewer: ```suggestion

```

=== Impact of Worker Count

As you deploy more functions or add workers per function, the per-worker quota decreases proportionally.

> A smaller per-worker quota can affect performance if individual functions require substantial memory for processing.

Reviewer: ```suggestion

```

== Queue Sizing and GC Behavior

> The per-worker quota directly influences 2 key aspects of Eventing runtime behavior.

Reviewer: ```suggestion

```

> === Worker Queue Sizing

Reviewer: ```suggestion
== Worker Queue Sizing
```

> Each worker maintains a bounded input queue that receives changes to documents from the Database Change Protocol (xref:server:learn:clusters-and-availability/intra-cluster-replication.adoc#database-change-protocol[DCP]) stream. 

Reviewer: ```suggestion
Each worker maintains a bounded input queue that receives changes to documents from the xref:server:learn:clusters-and-availability/intra-cluster-replication.adoc#database-change-protocol[Database Change Protocol (DCP)] stream. 
```

* The per-worker quota determines the maximum size of this queue.
* When the queue exceeds this size, Eventing throttles the xref:server:learn:clusters-and-availability/intra-cluster-replication.adoc#database-change-protocol[DCP] stream to prevent unbounded memory growth.

> * Throttling temporarily pauses mutation flow, ensuring workers do not become overwhelmed when processing lags behind the mutation rate. 

Reviewer: ```suggestion
The per-worker quota determines the maximum size of this queue.

When the queue exceeds this size, the Eventing Service throttles the DCP stream to prevent unbounded memory growth.

Throttling temporarily pauses the mutation flow, ensuring workers do not become overwhelmed when processing lags behind the mutation rate. 
```

> === Garbage Collection (GC) Triggering

Reviewer: ```suggestion
[#gc-trigger]
== Garbage Collection (GC) Triggering
```

* GC frees memory occupied by unreachable objects that are no longer in use.
* GC reclaims memory only from objects that are no longer in use. 
Memory held by live objects in the JavaScript runtime heap remains allocated, so total usage may exceed the per-worker quota.

NOTE: When memory usage approaches the per-worker quota limit, the JavaScript runtime may trigger garbage collection to reclaim memory. 
However, the garbage collector may delay the stop-the-world collection because it's optimized for throughput. 
As a result, the runtime may not reclaim memory without delay, which can temporarily affect Eventing memory consumption.

> The service invokes a stop-the-world JavaScript runtime GC to reduce memory inside the JavaScript runtime isolate running each function.

Reviewer: ```suggestion
The Eventing Service's garbage collection (GC) frees memory occupied by unreachable objects that are no longer in use by Eventing Functions.

Garbage collection reclaims memory only from objects that are no longer in use. 
Memory held by live objects in the JavaScript runtime heap remains allocated, so total usage may exceed the per-worker quota.

When memory usage approaches the per-worker quota limit, the JavaScript runtime may trigger garbage collection to reclaim memory. 

The garbage collector may delay the stop-the-world collection because it's optimized for throughput. 

The runtime might not reclaim memory without a delay, which can temporarily affect the Eventing Service's memory consumption.

The Eventing Service invokes a stop-the-world JavaScript runtime garbage collection to reduce memory inside the JavaScript runtime isolate running each Eventing Function.
```

== Limitations and Considerations
The Eventing Service memory quota has important limitations that users must understand to prevent operational issues:

* Not a hard limit: The quota does not cap total memory usage. 
Live objects in JavaScript runtime heaps may persist, so memory usage can exceed the quota.
* No OS-level enforcement outside cgroups: In VM or bare-metal deployments, the operating system does not enforce memory limits.
* Worker sizing impacts memory: Configuring an excessive number of workers or running memory-intensive functions can increase memory pressure and affect node stability.

> * Quota applies per worker, not per function: Each worker receives an equal share of the total quota, regardless of individual function memory requirements.

Reviewer: ```suggestion
== Limitations and Considerations

To prevent operational issues while using the Eventing Service, keep the following in mind while configuring the memory quota:

* <<no-hard-limit,>>
* <<no-enforce-c-groups,>>
* <<worker-sizing-memory,>>
* <<quota-per-worker,>>
```

> === Not a Hard Limit

Reviewer: ```suggestion
[#no-hard-limit]
=== Not a Hard Limit
```

The Eventing Service memory quota does not restrict total memory usage. 

JavaScript runtime heaps may still contain live (non-garbage) objects that:

* GC cannot reclaim.
* The quota does not account.
* Continue to consume memory beyond the configured limit.

> As a result, setting a memory quota does not guarantee that Eventing functions remain within that limit.

Reviewer: ```suggestion
* Garbage collection cannot reclaim.
* The quota does not include.
* Continue to consume memory beyond the configured limit.

As a result, setting a memory quota does not guarantee that Eventing Functions remain within that limit.
```

> === No OS-Level Enforcement Outside cgroups

Reviewer: ```suggestion
[#no-enforce-c-groups]
=== No OS-Level Enforcement Outside cgroups
```

> In non-containerized environments, the operating system does not enforce the memory quota.

Reviewer: The operating system? Or Couchbase Server/the Eventing Service?

Author: Need to get this verified by the SME

Author: Spoke to the SME and have changed the title of this section, added new content, and modified this section.

* Users must monitor Eventing memory usage using system tools.
* Configuring an excessive number of workers can lead to memory exhaustion.
* High-memory-consuming functions can affect overall system stability.

> * Eventing does not automatically stop or throttle execution when total memory exceeds the quota. 

Reviewer: ```suggestion
You must monitor the Eventing Service's memory usage using system tools.

If you configure too many workers for your Eventing Functions, you can cause memory exhaustion in your cluster. 
High memory-consuming Eventing Functions can affect your overall system stability.

Again, the Eventing Service does not automatically stop or throttle execution when total memory exceeds the quota. 
```

=== Memory Isolation

The memory quota does not isolate memory usage between Eventing functions.

* All workers draw from the same total memory quota.
* A memory-intensive function can reduce the memory available to other functions.
* Eventing does not support per-function memory limits or reservations.

== Best Practices
Follow these recommendations to effectively manage Eventing memory usage effectively and avoid operational issues.

=== Monitor Memory Usage

* Use system-level monitoring tools to track actual memory consumption.
* Monitor memory usage for the Eventing process and individual workers.
* Configure alerts for memory thresholds well below system limits.

=== Configure Worker Count Appropriately

* Consider the per-worker quota when determining the number of workers.
* Avoid over-provisioning workers when Eventing functions require substantial memory.

> * Balance the worker count against available system memory.

Reviewer: ```suggestion
[#worker-sizing-memory]
=== Worker Sizing Impacts Memory

When configuring the memory quota for the Eventing Service, consider the per-worker quota when determining the number of workers.
Avoid over-provisioning workers when Eventing Functions require substantial memory.

For best results, try to balance your worker count against your available memory. 

[#quota-per-worker]
=== Quota Applies Per Worker

The memory quota does not isolate memory usage between Eventing Functions.
All workers draw from the same total memory quota.

A memory-intensive function can reduce the memory available to other Eventing Functions.

Eventing does not support per-function memory limits or reservations.

== Best Practices

Follow these recommendations to effectively manage Eventing memory usage: 

* <<monitor,>>
* <<test,>>
* <<use-cgroups,>>
* <<size,>>
* <<review,>>

[#monitor]
=== Monitor Memory Usage

* Use system-level monitoring tools to track actual memory consumption.
* Monitor memory usage for the Eventing process and individual workers.
* Configure alerts for memory thresholds well below system limits.
```

=== Test in Representative Environments

* Test Eventing functions under production-like load conditions.
* Measure actual memory consumption during peak processing.
* Validate that memory usage remains within acceptable limits.

=== Use cgroups for Enforcement

* Consider containerized deployments or cgroup-based memory limits in production environments.
* Container platforms such as Docker and Kubernetes provide hard memory limits.
* cgroups enforce OS-level memory restrictions.
* This approach adds a safety layer beyond the Eventing Service memory quota.

=== Size the Memory Quota Appropriately

* Set the Eventing memory quota based on expected worker count and function complexity.
* Allow additional headroom on top of the 256 MB default for production workloads.
* Account for peak processing scenarios, not just average load.

=== Review Function Memory Efficiency

* Optimize JavaScript code to minimize memory allocation.
* Avoid accumulating large data structures in function scope.
* Release object references when data is no longer needed.

> * Profile functions to identify memory-intensive operations.

Reviewer: ```suggestion
[#test]
=== Test in Representative Environments

* Test Eventing Functions under production-like load conditions.
* Measure actual memory consumption during peak processing.
* Validate that memory usage remains within acceptable limits.

[#use-cgroups]
=== Use cgroups for Enforcement

* Consider containerized deployments or cgroup-based memory limits in production environments.
* Container platforms such as Docker and Kubernetes provide hard memory limits.
* cgroups enforce OS-level memory restrictions.
This approach adds a safety layer beyond the Eventing Service memory quota.

[#size]
=== Size the Memory Quota Appropriately

* Set the Eventing Service memory quota based on expected worker count and Eventing Function complexity.
* Allow additional headroom on top of the 256{nbsp}MB default for production workloads.
* Account for peak processing scenarios, not just average load.

[#review]
=== Review Function Memory Efficiency

* Optimize JavaScript code to minimize memory allocation.
* Avoid accumulating large data structures in the function scope.
* Release object references when data is no longer needed.
* Profile Eventing Functions to identify memory-intensive operations.
```

// Learn how to provide links to other doc pages and provide references to pages like Eventing Service Overview, Eventing Service Settings, and Eventing Function Examples. 
// == Related Links 
// * xref:eventing-service-overview.adoc[Eventing Service Overview]
// * xref:eventing-service-settings.adoc[Eventing Service Settings]

> // * 

Reviewer: Clean this up, please. Don't just leave this commented out with this kind of comment.
